scenario/smartsearch.go

```diff
@@ -0,0 +1,200 @@
+		suggestions = append(suggestions, quote(suggestion.Value))
+	}
+	if len(suggestions) > 0 {
+		logEntry.LogDebugf("search suggestions: %s", strings.Join(suggestions, ","))
+	}
+	for _, sga := range searchResult.SearchGroupArray {
+		searchTermsMatched := []string{}
+		for _, stIndex := range sga.SearchTermsMatched {
+			searchTermsMatched = append(searchTermsMatched, quote(searchResult.SearchTerms[stIndex]))
+		}
+		for _, item := range sga.Items {
+			matches := []string{}
+			for _, match := range item.ItemMatches {
+				matches = append(matches, quote(match.Text))
+			}
+			logEntry.LogDebugf(`search result for term%s %s in %s %s: %s`,
+				ternary(len(searchTermsMatched) != 1, "s", ""), strings.Join(searchTermsMatched, ","),
+				item.ItemType, quote(item.Identifier), strings.Join(matches, ","))
+		}
+	}
+}
+
+var spaceRegex = regexp.MustCompile(`\s+`)
+
+func parseSearchTerms(s string) []string {
+	searchTerms := []string{}
+	sb := &strings.Builder{}
+	quoted := false
+	escaped := false
+	addSearchTerm := func() {
+		defer sb.Reset()
+		searchTerm := sb.String()
+		searchTerm = strings.TrimSpace(searchTerm)
+		searchTerm = spaceRegex.ReplaceAllString(searchTerm, " ")
+		if searchTerm != "" {
+			searchTerms = append(searchTerms, searchTerm)
+		}
+	}
+	for _, r := range s {
+		if escaped {
+			sb.WriteRune(r)
+		} else if r == '\\' {
+			escaped = true
+			continue
+		} else if r == '"' {
+			quoted = !quoted
+		} else if !quoted && unicode.IsSpace(r) {
+			addSearchTerm()
+		} else {
+			sb.WriteRune(r)
+		}
+		escaped = false
+	}
+	addSearchTerm()
+	return searchTerms
+}
+
+// Execute implements ActionSettings interface
+func (settings SmartSearchSettings) Execute(sessionState *session.State, actionState *action.State, connection *connection.ConnectionSettings, label string, reset func()) {
+	uplink := sessionState.Connection.Sense()
+	if uplink.CurrentApp == nil {
+		actionState.AddErrors(errors.New("not connected to app"))
+		return
+	}
+	doc := uplink.CurrentApp.Doc
+	searchTerms := settings.searchTerms
+	sessionState.LogEntry.LogDebugf(`search terms: "%s"`, strings.Join(searchTerms, `","`))
+
+	var searchResult *enigma.SearchResult
+	var searchSuggestionResult *enigma.SearchSuggestionResult
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		err := sessionState.SendRequest(actionState, func(ctx context.Context) error {
+			var err error
+			searchSuggestionResult, err = doc.SearchSuggest(
+				ctx,
+				&enigma.SearchCombinationOptions{},
+				searchTerms,
+			)
+			return err
+		})
+		if err != nil {
+			actionState.AddErrors(err)
+		}
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		err := sessionState.SendRequest(actionState, func(ctx context.Context) error {
+			var err error
+			searchResult, err = doc.SearchResults(
+				ctx,
+				&enigma.SearchCombinationOptions{
+					Context:      "CurrentSelections",
+					CharEncoding: "Utf16",
+				},
+				searchTerms,
```

Comment:
So this will always send all search times as one request?

Reply:
Yes, then you get a one list item each in the response for all combinations of terms.